immuni_analytics/celery/authorization/tasks/authorize_analytics_token.py

```diff
@@ -0,0 +1,163 @@
+                bit1=device_check_data.bit1,
+                last_update_time=device_check_data.last_update_time,
+            ),
+        )
+        await _blacklist_device(device_token)
+
+
+async def _second_step(device_token: str) -> None:
+    """
+    Fetch the DeviceCheck bits and ensure the configuration is expected.
+    If it is, set the bits to (0,1) and return.
+    If not, blacklist the device.
+
+    :raises: DiscardAnalyticsTokenException
+    """
+    device_check_data = await fetch_device_check_bits(device_token)
+    if not device_check_data.is_default_configuration_compliant:
+        _LOGGER.warning(
+            "Found token not default configuration compliant in second step.",
+            extra=dict(
+                env=config.ENV.value,
+                bit0=device_check_data.bit0,
+                bit1=device_check_data.bit1,
+                last_update_time=device_check_data.last_update_time,
+            ),
+        )
+        await _blacklist_device(device_token)
+
+    await set_device_check_bits(device_token, bit0=True, bit1=False)
+
+
+async def _third_step(device_token: str) -> None:
+    """
+    Fetch the DeviceCheck bits and ensure the configuration is expected.
+    If it is, set the bits to (0,0) and return.
+    If not, blacklist the device.
+
+    :raises: DiscardAnalyticsTokenException
+    """
+    device_check_data = await fetch_device_check_bits(device_token)
+    if not device_check_data.is_authorized_configuration_compliant:
+        _LOGGER.warning(
+            "Found token not authorization configuration compliant in third step.",
+            extra=dict(
+                env=config.ENV.value,
+                bit0=device_check_data.bit0,
+                bit1=device_check_data.bit1,
+                last_update_time=device_check_data.last_update_time,
+            ),
+        )
+        await _blacklist_device(device_token)
+
+    await set_device_check_bits(device_token, bit0=False, bit1=False)
+
+
+async def _blacklist_device(device_token: str) -> None:
+    """
+    Set the both DeviceCheck bits to True, a configuration that marks blacklisted devices.
+
+    :raises: DiscardAnalyticsTokenException
+    """
+    if config.ENV == Environment.RELEASE:
+        await set_device_check_bits(device_token, bit0=True, bit1=True)
+
+    raise DiscardAnalyticsTokenException()
+
+
+async def _add_analytics_token_to_redis(analytics_token: str) -> None:
+    """
+    Add the analytics token to the sets corresponding to the current and the next months.
+    """
+    pipe = managers.analytics_redis.pipeline()
+    pipe.sadd(get_authorized_tokens_redis_key_current_month(with_exposure=True), analytics_token)
+    pipe.sadd(get_authorized_tokens_redis_key_current_month(with_exposure=False), analytics_token)
+    pipe.sadd(get_authorized_tokens_redis_key_next_month(with_exposure=True), analytics_token)
+    pipe.sadd(get_authorized_tokens_redis_key_next_month(with_exposure=False), analytics_token)
```

Comment:
Thinking about how we perform the check 

    await managers.analytics_redis.sismember(
        get_authorized_tokens_redis_key_current_month(with_exposure=True), analytics_token
    ) or await managers.analytics_redis.sismember(
        get_authorized_tokens_redis_key_current_month(with_exposure=False), analytics_token
    ):

We could refactor this part so that it only requires one single check. Maybe using a hash table that contains a dictionary like 

```
{
    <month>: <exposure>[0, 1] (this is both with and without)
}
```

This would allow you to perform the check by fetching a single key (even better performances than a pipeline)